ikarus/io/vtkwriter.hh

```diff
@@ -0,0 +1,182 @@
+// SPDX-FileCopyrightText: 2021-2024 The Ikarus Developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+/**
+ * \file vtkwriter.hh
+ * \brief Ikarus VTK Writer for finite element results
+ * \ingroup io
+ *
+ */
+
+#pragma once
+
+#include <dune/functions/gridfunctions/discreteglobalbasisfunction.hh>
+#include <dune/vtk/vtkwriter.hh>
+#include <dune/vtk/writers/unstructuredgridwriter.hh>
+
+#include <ikarus/io/resultfunction.hh>
+#include <ikarus/utils/concepts.hh>
+
+namespace Ikarus::Concepts {
+
+// adapted from /dune/dune-vtk/dune/vtk/utility/concepts.hh
+template <class DC>
+concept IsDataCollector = requires(DC dc) {
+  typename DC::GridView;
+  dc.update();
+  dc.numPoints();
+  dc.numCells();
+};
+} // namespace Ikarus::Concepts
+
+namespace Ikarus::Vtk {
+
+namespace Impl {
+  struct AsCellData
+  {
+  };
+  struct AsPointData
+  {
+  };
+  namespace Concepts {
+    template <typename DT>
+    concept DataType = std::is_same_v<DT, AsCellData> or std::is_same_v<DT, AsPointData>;
+  }
+} // namespace Impl
+
+template <typename AS, bool structured = false, typename DC = void>
+requires(Concepts::FlatAssembler<AS> and (std::is_same_v<DC, void> or Concepts::IsDataCollector<DC>))
+class Writer
+{
+public:
+  using Assembler     = AS;
+  using GridView      = typename Assembler::GridView;
+  using FERequirement = typename Assembler::FERequirement;
+  using FEContainer   = typename Assembler::FEContainerType;
+  using FEType        = typename std::remove_cvref_t<FEContainer>::value_type;
+
+  // We are using the provided DataCollector, but if none was provided we are using the default ones from dune-vtk
+  using DataCollector = std::conditional_t<std::is_same_v<DC, void>,
```

Comment:
I would suggest we have a struct defined outside:
```cpp
template<typename GV>
struct DefaultDataCollector
{};
```
which can be then be specialized for different gridviews as

```cpp
template<...>
struct DefaultDataCollector<typename Dune::YaspGrid<....>::GridView>
{
 using Type= Dune::Vtk::YaspDataCollector<typename Dune::YaspGrid<....>::GridView>;
};
```

```cpp
template <typename AS, bool structured = false, typename DC = DefaultDataCollector<Assembler::GridView>>
requires(Concepts::FlatAssembler<AS> and (std::is_same_v<DC, void> or Concepts::IsDataCollector<DC>))
class Writer
```

Then we can also have specialization in `dune-iga` for trimmed grids.
With this customization point, we do not often have to touch this template argument.

Reply:
How does this specialization work her? 

```cpp
template <int dim, typename Coordinates>
struct DefaultDataCollector<typename Dune::YaspGrid<dim, Coordinates>::LeafGridView> {
  using Type = Dune::Vtk::YaspDataCollector<typename Dune::YaspGrid<dim, Coordinates>::LeafGridView>;
};
```
This doesnt work as i can not deduce dim and Coordinates
Also I think that has to be spacialized then for `Leaf`  and `LevelGridView`